### ISSUE TITLE
speed improvements

### ISSUE THREAD

**Issue by duartemolha** [genome/bam-readcount#48]
Hi... thank you for this tool ... it is very helpful

But... I was wondering... is there a way we could maybe speed this up by using threads.

I have been trying to use it to calculate quite a few regions and it is very slow.

I could split by regions into several files and run them in parallel but if there was a way of maybe doing that directly on the software by using threads that would be much better.

I have a server with more than enough memory and cpu ... so it would be great if we could specify this

**Comment by duartemolha** [comment]
Up?



[image: --]
Duarte Molha
[image: https://]

**Comment by daz10000** [comment]
I recently tested bam-readcounts and love the tool, but it's one of the slowest steps in my workflow and I noticed it's only using a single CPU (running on machines with 16-32 CPUS).  It seems like it should be highly parallel, but maybe IO is a bigger bottleneck?  I have considered just splitting the input files N ways and running N copies of bam-readcounts as a first benchmark, but curious if you have any thoughts on speeding it up?  I'd be happy to take a look at a threading implementation myself - do you have any instincts around where the CPU is most intense? ( I can parallelize at a high level but might be more efficient to get to an inner loop).  Any thoughts appreciated - and if there is already a parallel implementation I missed, apologies.

**Comment by apldx** [comment]
We haven't done any benchmarking that I know of, and I usually max out at 10,000 or 20,000 loci (we are usually interested in somatic variants) which isn't much, nor performance analysis to identify where time is being spent. Under the hood, we're using the htslib pileup engine but we may be adding some inefficiency in putting together the statistics. Would be happy to keep the conversation going and keen to hear your results if you do some benchmarking!

**Comment by daz10000** [comment]
Out of curiosity, how long does a typical run take?  I might have an unusual data set (large number of sites - sometimes 1.5M), but I'm seeing highly variable runtimes, and for the largest sets I saw one take 11 hours (though more often in the 4-5 hr range).  The other strangeness is apparently very different runtimes across different CPUs (slow on ARM, fast on intel), even though the ARM chip is generally quicker for everything else.  At one point I watched the rate of line generation and it just crawled in some places, but was faster in others - might be chromosome size or depth of reads affecting it.  Hard to say without profiling.

One suspicion looking at the top loop (I haven't measured yet, or played with the code, but wanted to talk first! and probably profile before touching code- the wisdom of age),   is that it appears to be fetching the reference sequence for each variant (let me know if I'm misreading here).  If the CPU cache were large enough, this might be harmless, but if it's not cache friendly, I could see how this would create a lot of work.   I was also thinking inserting threading here to do blocks of variants would help if it's primarily CPU limited.    Watching it run, it pegs one CPU at 100% (out of 16 on the machine), so it could benefit from more CPUs working in parallel.

I will try to get some profiling data, but any thoughts you have appreciated. I'm looking at this code block right now and trying to see what it does per variant.  Am I right in assuming it's fetching the reference sequence for a chromosome each row, even if the chromosome is same as the previous row?

![image](https://github.com/genome/bam-readcount/assets/6342645/76f507bc-be82-4a8a-9219-676f4f2850c2)

**Comment by daz10000** [comment]
One practical question - I haven't had to deal with CMAKE for a while (honestly I find it super annoying - too much hidden magic).  Do you have tips for building with `-pg` for linker and compiler.  I have tried 3 things so far that didn't change the build and running out of ideas.

**Comment by daz10000** [comment]
Still not loving CMAKE but got profiling built.  Here is output from running 5k sites (I don't know if there is something horribly non linear that happens in some parts of the genome that this won't capture - I should maybe pick 5k random sites across the genome rather than the first 5k sites), but it's a start.  Fairly simple answer on where the time goes.  It's all in zlib.  I'm processing this file many different ways and the other pipelines aren't choking as badly which is odd.  I'll make some more measurements and try some random sites across the genome maybe.  I don't imagine there's any problem with the built from source zlib, but will also try swapping out the system package zlib (same version) to see if it's optimized in any strange way,

```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total
 time   seconds   seconds    calls  us/call  us/call  name
 64.29    328.75   328.75                             inflate_fast
 24.54    454.24   125.49                             crc32_little
  3.65    472.93    18.69                             overlap_remove
  1.84    482.32     9.39                             bam_read1
  1.19    488.41     6.09                             inflate_table
  0.90    493.03     4.62 24844950     0.19     0.19  fetch_func(bam1_t const*, void*)
  0.66    496.39     3.36                             bgzf_read
  0.62    499.55     3.16                             bam_aux_get
  0.59    502.58     3.03                             inflate
  0.30    504.10     1.52                             bam_tag2cigar
  0.20    505.10     1.00                             hts_itr_next
  0.16    505.92     0.82   914326     0.90     1.63  pileup_func(unsigned int, unsigned int, int, bam_pileup1_t const*, void*)
  0.12    506.53     0.61                             sam_readrec
  0.12    507.13     0.60                             fai_retrieve
  0.12    507.73     0.60                             _init
  0.11    508.30     0.57                             sam_read1
  0.09    508.78     0.48                             bam_endpos
  0.06    509.11     0.33                             sam_dispatcher_write
  0.06    509.41     0.30 13400585     0.02     0.02  aux_zm_t::from_string(char const*)
  0.05    509.67     0.26                             bgzf_getc
  0.04    509.87     0.20 13400585     0.01     0.04  BasicStat::process_read(bam_pileup1_t const*)
  0.03    510.03     0.16 146292824     0.00     0.00  kh_resize_s(kh_s_s*, unsigned int)
  0.03    510.19     0.16                             bam_plp_destroy
  0.03    510.33     0.14                             samfetch
  0.03    510.47     0.14                             bgzf_read_block
  0.02    510.58     0.11                             bam_plbuf_push
  0.02    510.67     0.09                             bam_plp64_next
  0.02    510.76     0.09                             bam_plp_push
  0.02    510.85     0.09                             sam_realloc_bam_data
  0.02    510.93     0.08                             inflateInit2_
  0.02    511.01     0.08                             bam_copy1
  0.01    511.08     0.07                             bam_aux_append
```

**Comment by daz10000** [comment]
First experiment (swapping out the vendor built in zlib) does completely change the profiling data.  I didn't think to do a wallclock time :( - need to check absolute time,  but this time the dominant calls are not just unpacking the bam file [ caveat - since I'm using a system libz now, it wouldn't be profiled explicitly but just the cumulative seconds here is 47 vs over 500 above].  

Edit: studying the wall time,  I don't think swapping out for the system libz made things faster,  it probably just obscured the time spent

```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total
 time   seconds   seconds    calls  us/call  us/call  name
 39.31     18.47    18.47                             overlap_remove
 19.54     27.65     9.18                             bam_read1
 10.49     32.58     4.93 24844950     0.20     0.20  fetch_func(bam1_t const*, void*)
  6.98     35.86     3.28                             bam_aux_get
  6.65     38.99     3.13                             bgzf_read
  2.79     40.30     1.31                             bam_tag2cigar
  1.98     41.23     0.93                             hts_itr_next
  1.57     41.97     0.74   914326     0.81     1.59  pileup_func(unsigned int, unsigned int, int, bam_pileup1_t const*, void*)
  1.40     42.63     0.66                             sam_readrec
  1.19     43.19     0.56                             fai_retrieve
  1.09     43.70     0.51                             _init
  1.06     44.20     0.50                             bam_endpos
  0.85     44.60     0.40                             sam_read1
  0.64     44.90     0.30                             bgzf_getc
  0.62     45.19     0.29 13400585     0.02     0.04  BasicStat::process_read(bam_pileup1_t const*)
  0.60     45.47     0.28                             sam_dispatcher_write
  0.53     45.72     0.25                             samfetch
  0.51     45.96     0.24 13400585     0.02     0.02  aux_zm_t::from_string(char const*)
  0.38     46.14     0.18 146292824     0.00     0.00  kh_resize_s(kh_s_s*, unsigned int)
  0.30     46.28     0.14                             bam_plp_destroy
  0.26     46.40     0.12                             bgzf_read_block
  0.19     46.49     0.09                             bam_aux_append
  0.17     46.57     0.08                             bam_plp_push
  0.13     46.63     0.06                             sam_realloc_bam_data
  0.11     46.68     0.05                             bam_copy1
  0.11     46.73     0.05                             bam_plbuf_push
  0.11     46.78     0.05                             bam_plp64_next
  0.09     46.82     0.04                             hread2
  0.06     46.85     0.03                             bgzf_uncompress
  0.04     46.87     0.02                             fd_read
  0.04     46.89     0.02                             fd_write
  0.04     46.91     0.02                             refill_buffer
```

**Comment by daz10000** [comment]
Assuming it's mainly bam decompression that's taking all the time,  there is a faster `libz` implementation covered nicely [here](https://learn.arm.com/learning-paths/servers-and-cloud-computing/zlib/setup/) from cloudflare.  Simply swapping that in should help.  I've benchmarked my 50k site test set (with a 9.8 Gb bam file).

## System zlib (no crc32 acceleration on arm)

```
real    9m10.026s
user    8m59.932s
sys     0m10.070s
```


## Cloudflare zlib (uses accelerated instruction sets)
```
real    5m16.411s
user    5m6.175s
sys     0m10.220s
```

So about half the time. (NB: this is a pita to build - I had to edit the link.txt file in the cmake build tree, swap out the vendor libz , link against the system version with `-lz` which links dyanamically then use the resulting binary with the cloudflare zlib).  The default bam-readcount build made a statically linked libz that can't be substituted easily.

## Conclusions

* bam-readcount is mostly limited by the speed of processing (specifically uncompression of the input bam file)
* On my arm system, the native libz (and the one shipped with bam-readcounts) has no acceleration
* Installing the `cloudflare-libz` library roughly doubles the speed
* The vanilla version is processing only about 9 sites per second on my data set :(

I'm inferring from all of this too that a parallel version of the code (either dumb , splitting sites up and running many copies or inline with threading) could greatly benefit performance, since it's apparently not IO bound but CPU decompressing.

Would love to hear your thoughts.

## Addendum

Running two jobs on the same data set in parallel (dumb parallelism model) does roughly halve the runtime (uses two cpus!) but you have to carefully divide up the work.  Naively splitting my test set into two resulted in one job that took 10 seconds and another that still took 5 minutes.  The work seems to be very dependent on the depth of reads not surprisingly,  so you probably want to evenly chop up your sites in a rotating manner if you try this at scale.  Threading built into the code would be much more elegant.


```
real    2m38.552s
user    2m33.391s
sys     0m5.153s

real    2m38.257s
user    2m33.367s
sys     0m4.880s
```

**Comment by daz10000** [comment]
I'm doing an expanded 'dumb parallelism' experiment with 16 way input file splitting (interleaved as described above) and I can confirm on my 1.5M site test set, that it remains CPU bound.

![image](https://github.com/genome/bam-readcount/assets/6342645/676b18c1-6d1c-4df8-a534-384655a56431)

That whole set finished in almost exactly 20 mins with the 16 cpus (for 1.5M sites, 2.9Gb bam file).  I think that's fast enough for me.  Obviously if there are cleaner ways of getting this (I have to pre and post process the data files here a lot), I'm interested in hearing them.  It does mean that dropping threading into that inner loop would probably get the same user experience without all this external juggling.

The original 5k site test set I benchmarked with, runs in 20 seconds with the final setup vs 10 minutes when I started.

**Comment by apldx** [comment]
Thanks for the detailed writeups! Catching up with your results now. One thing that comes to mind is that we could make it easier for users to build `bam-readcount` against shared system libraries instead of the bundled versions. The idea was to make it self-contained but we lose the flexibility of dropping in something like the Cloudflare `zlib`.

**Comment by daz10000** [comment]
thanks!  yes the bundling made it quite tricky to slip in the alt library and even to tell which one the binary was using.  It’s tricky either way. I get the ease argument.  I’m used to just needing to install half a dozen prerequisites before I build something from C code anyway.  It would be really nice to have threading built into the inner loop - that’s what made the biggest difference in the end, though I was surprised home time went into decompressionAm 11/21/23 um 8:14 AM schrieb Ajay Khanna ***@***.***>:﻿
Thanks for the detailed writeups! Catching up with your results now. One thing that comes to mind is that we could make it easier for users to build bam-readcount against shared system libraries instead of the bundled versions. The idea was to make it self-contained but we lose the flexibility of dropping in something like the Cloudflare zlib.
